python/graphstorm/dataloading/dataloading.py

```diff
@@ -917,3 +921,170 @@ def __iter__(self):
+
+####################### Distillation #############################
+
+class DataProvider:
+    r"""Data Provider. Combines a file sampler and a dataloader generator,
+    and streamingly provides an iterable over a dataset.
+
+    Parameters:
+    ----------
+        dataloader_generator : DataloaderGenerator: 
+            A dataloader generator
+            Generates dataloader based on given file path.
+        dataset_path str : 
+            Path to the data files.
+        shuffle : bool
+            Set to ``True`` to have the files reshuffled at every epoch (default: ``False``).
+        infinite : bool
+            Set to ``True`` to have the files sampled infinitely (default: ``False``).
+        local_rank : int
+            Local rank for the trainer (default: ``-1``).
+        is_train : bool
+            Set to ``True`` if the provider is for training set (default: ``True``).
+    """
+
+    def __init__(
+        self,
+        dataloader_generator,
+        dataset_path,
+        shuffle=False,
+        infinite=False,
+        local_rank=-1,
+        world_size=1,
+        is_train=True,
+    ):
+        super().__init__()
+        # TODO (HZ): implement prefetch to put files into self.dataset_future
+        self.dataset_future = []
+        
+        if dataset_path is not None:
+            if not isinstance(dataset_path, str):
+                raise TypeError(
+                    "dataset_path should be a str, but got "
+                    "dataset_path={}".format(dataset_path)
+                )
+            file_sampler = DistributedFileSampler(
+                dataset_path=dataset_path,
+                shuffle=shuffle,
+                infinite=infinite,
+                local_rank=local_rank,
+                world_size=world_size,
+                is_train=is_train,
+            )
+        else:
+            raise ValueError(
+                    "dataset_path needs to be specified"
+                )
+
+        self.file_sampler = file_sampler
+        self.file_sampler_iter = iter(self.file_sampler)
+
+        self.dataloader_generator = dataloader_generator
+
+        self.dataloader = None
+        self.data_file = None
+
+        print (f"DataProvider - Initialization: num_files = {len(file_sampler)}")
+
+    def __del__(self):
+        for data_file, data_job in self.dataset_future:
+            if data_job:
+                path, sample_count, _ = data_job.result(timeout=None)
+                shutil.rmtree(path)
+
+    def get_shard_name(self):
+        """ Return shard name.
+        """
+        return self.data_file
+
+    def get_shard(self):
+        """ Get dataloader from a data file.
+        """
+        dataloader = None
+        sample_count = 0
+
+        if len(self.dataset_future) == 0:
+            data_file = next(self.file_sampler_iter)
+        else:
+            # when DataProvider does prefetches
+            data_file, _ = self.dataset_future.pop(0)
```

Comment:
Where do you put the data into dataset_future?

Reply:
This is ideally to be used by prefetch. Since it's not implemented yet i'll remove.